File: 38.count-and-say.go
import "strconv"

/*
 * @lc app=leetcode id=38 lang=golang
 *
 * [38] Count and Say
 */

// @lc code=start

// TODO understand this shit
func countAndSay(n int) string {
	if n == 1 {
		return "1"
	}
	result := "1"
	for i := 1; i < n; i++ {
		temp := ""
		j := 0
		for j < len(result) {
			count := 1
			currentChar := result[j]

			for j < len(result)-1 && result[j] == result[j+1] {
				count++
				j++
			}
			j++

			temp += strconv.Itoa(count) + string([]byte{currentChar})
		}
		result = temp
	}
	return result
}

// @lc code=end

File: 38.count-and-say.cpp
/*
 * @lc app=leetcode id=38 lang=cpp
 *
 * [38] Count and Say
 */

// @lc code=start
class Solution {
public:
    string countAndSay(int n) {
        if (n == 0) return "";

        string result = "1";
        for (int i = 0; i < n - 1; i++) {
            string temp = "";
            for (int j = 0; j < result.size(); j++) {
                int count = 1;
                while ((j + 1 < result.size()) && (result[j] == result[j + 1])) {
                    count++;
                    j++;
                }
                temp += to_string(count) + result[j];
            }
            result = temp;
        }

        return result;
    }
};
// @lc code=end

